src/script/miniscript.h

```diff
@@ -0,0 +1,1563 @@
+                case NodeType::OR_C: return std::move(subs[0]) + (CScript() << OP_NOTIF) + std::move(subs[1]) + (CScript() << OP_ENDIF);
+                case NodeType::OR_I: return (CScript() << OP_IF) + std::move(subs[0]) + (CScript() << OP_ELSE) + std::move(subs[1]) + (CScript() << OP_ENDIF);
+                case NodeType::ANDOR: return std::move(subs[0]) + (CScript() << OP_NOTIF) + std::move(subs[2]) + (CScript() << OP_ELSE) + std::move(subs[1]) + (CScript() << OP_ENDIF);
+                case NodeType::MULTI: {
+                    CScript script = CScript() << node.k;
+                    for (const auto& key : node.keys) {
+                        script << ctx.ToPKBytes(key);
+                    }
+                    return std::move(script) << node.keys.size() << (verify ? OP_CHECKMULTISIGVERIFY : OP_CHECKMULTISIG);
+                }
+                case NodeType::THRESH: {
+                    CScript script = std::move(subs[0]);
+                    for (size_t i = 1; i < subs.size(); ++i) {
+                        script = (std::move(script) + std::move(subs[i])) << OP_ADD;
+                    }
+                    return std::move(script) << node.k << (verify ? OP_EQUALVERIFY : OP_EQUAL);
+                }
+            }
+            assert(false);
+            return {};
+        };
+        return TreeEval<CScript>(false, downfn, upfn);
+    }
+
+    template<typename CTx>
+    bool ToString(const CTx& ctx, std::string& ret) const {
+        // To construct the std::string representation for a Miniscript object, we use
+        // the TreeEvalMaybe algorithm. The State is a boolean: whether the parent node is a
+        // wrapper. If so, non-wrapper expressions must be prefixed with a ":".
+        auto downfn = [](bool, const Node& node, size_t) {
+            return (node.nodetype == NodeType::WRAP_A || node.nodetype == NodeType::WRAP_S ||
+                    node.nodetype == NodeType::WRAP_D || node.nodetype == NodeType::WRAP_V ||
+                    node.nodetype == NodeType::WRAP_J || node.nodetype == NodeType::WRAP_N ||
+                    node.nodetype == NodeType::WRAP_C ||
+                    (node.nodetype == NodeType::AND_V && node.subs[1]->nodetype == NodeType::JUST_1) ||
+                    (node.nodetype == NodeType::OR_I && node.subs[0]->nodetype == NodeType::JUST_0) ||
+                    (node.nodetype == NodeType::OR_I && node.subs[1]->nodetype == NodeType::JUST_0));
+        };
+        // The upward function computes for a node, given whether its parent is a wrapper,
+        // and the string representations of its child nodes, the string representation of the node.
+        auto upfn = [&ctx](bool wrapped, const Node& node, Span<std::string> subs) -> std::optional<std::string> {
+            std::string ret = wrapped ? ":" : "";
+            std::stringstream k_sstr;
```

Comment:
Would it compile if you remove this and use `::ToString(node.k)` instead?

Reply:
:man_facepalming: Yes. Done.